cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -142,6 +133,16 @@ impl fmt::Display for YaccGrammarErrorKind {
     }
 }
 
+impl YaccGrammarError {
+    pub fn span(&self) -> Span {
```

Comment:
I'm not sure about the existence of this method to be honest. At least it would have to be called `first_span` or `base_span` or similar? I think I would prefer not to privilege any of the spans unless we have a good use case for doing so beyond the test suite.

Reply:
I was thinking `primary_span`, if that is ok?

So, the reason for separating these is I wanted to send a patch to format strings for extra spans,
Currently I hard code a string such as "Additionally declared here", to `extra_spans`.

Further we don't see *all* the spans as being an error, for instance if there are 2 duplicates, only one of those
needs to be removed, so it seemed to me there is a semantic difference between the spans.  That is generally why I had kept them separate, so that in a future patch I could work towards improving a better reflection of this semantic difference.

Reply:
I'm thinking of some compilers where the "main" error is at line 10 but it will tell you "lines 7 and 14 are also related". I suppose we *could* differentiate "primary" spans (notice the plural!) from "secondary" spans but I'm not sure it's worth the complication: I think I'm fine having an ordered sequence of spans. It also means that if someone knows something special about one of the kinds, they might be able to interpret the sequence of spans in a clever way (whereas having methods to access them does implicitly place emphasis on the first entry).

Reply:
Ok, one thing to consider is that with `kind  now private, We don't have any mechanism to differentiate span order relative to kind, except by querying the `errs.len()`.
We could consider a hint like
```
/// Gives an indication of the error type 
enum YaccGrammarErrorType {
  DuplicationError,
  Error,
}

fn error_type(&self: YaccGrammarError) -> YaccGrammarErrorType { ... }
```

I'm a little uncomfortable interpreting the sequence of spans, without a mechanism in place to differentiate them.

Reply:
I like the documentation idea -- that gives us lots of flexibility but isn't too onerous!

Reply:
Should be fixed in https://github.com/softdevteam/grmtools/pull/322/commits/65d12bfc088d9a8d69cf7824da79daa22424666b

After this patch it might take me a bit to figure out if there is a nice way to avoid squashing this all into a single commit,
But i'm not sure.

Reply:
I now realise I had completely misunderstood your message -- mea culpa! What I *thought* you meant was that we would do:

```rust
enum YaccGrammarErrorKind {
   /// Only has 1 span
   IllegalInteger,
   /// First Span is the original, subsequent spans are duplicates.
   DuplicateRule,
   ...
}
```

and so on (hence my "documentation idea" comment).

I think your approach in https://github.com/softdevteam/grmtools/commit/65d12bfc088d9a8d69cf7824da79daa22424666b is a better approach. I suggest we push it a bit further by:

* `error_type` should be `spans_kind` returning a `SpansKind` which would be along the lines of your existing `enum`:
```rust
/// Gives an indication of the error type and interpretation of it's data.
pub enum SpansKind {
    /// The first span is the first occurrence, and a span for each subsequent occurrence.
    DuplicationError,
    /// Contains a single span at the site of the error.
    Error,
}
```
* *Perhaps* (this might not be a good idea!) we turn the `spans` method into `iter_spans` so that we don't have to expose the underlying `Vec` quite so obviously? But that might a bit annoying to use...

Reply:
Yeah `SpansKind` is a better name,

Just for clarification the reason against documenting via the `YaccGrammarErrorKind` variants is because they are now private.  So in practice if it is documented there it is hard leverage that documentation, or even make it visible.

These two things should be fixed in https://github.com/softdevteam/grmtools/pull/322/commits/d70d0357dbca29ba90223a52d6183b418dfb56d0 (rename), and https://github.com/softdevteam/grmtools/pull/322/commits/6ff70b3dc8488bcc2f4c8ad47198594824bea205 (use iter)

Reply:
I guess I should summarize that kind of long winded text (sorry)...

1. I kind of don't see the need outside of the testsuite to access kind, they generally just get used via conversion to `String`
2. There is some benefit to making them private in that we can add/remove without bumping semver.

I'm not really opposed to making them `pub` again though, it isn't important to me that these be private, just seemed like it was worth considering for the semver flexibility.

Reply:
Duh, I forgot the semver thing. Agreed, let's keep them private at least for the time being. I *think* one day we'll want to open them up but, hopefully, when someone actually needs that, the set of errors will have stabilised.